lib/wallet/src/Cardano/Wallet/Types/Read/Tx.hs

```diff
@@ -12,142 +11,36 @@ Copyright: © 2022 IOHK
-    Transaction type
--------------------------------------------------------------------------------}
 -- | Closed type family returning the ledger 'Tx' type for each known @era@.
-type family TxEra era where
-    TxEra ByronEra = Byron.ATxAux ()
-    TxEra ShelleyEra = Shelley.Tx (Api.ShelleyLedgerEra ShelleyEra)
-    TxEra AllegraEra = Shelley.Tx (Api.ShelleyLedgerEra AllegraEra)
-    TxEra MaryEra = Shelley.Tx (Api.ShelleyLedgerEra MaryEra)
-    TxEra AlonzoEra = Alonzo.ValidatedTx (Api.ShelleyLedgerEra AlonzoEra)
-    TxEra BabbageEra = Alonzo.ValidatedTx (Api.ShelleyLedgerEra BabbageEra)
-
-{- Note [CardanoEra]
-
-The `CardanoEra` from `Cardano.Api` enumerates the known Cardano eras.
-It's very convenient, because it's a GADT and
-brings a type level `era` in scope.
-
-This `era` is not immediately suitable for use with the ledger types,
-but the type family `Api.ShelleyLedgerEra` performs the conversion.
--}
-
--- | A transaction as it can be read from the Cardano mainnet.
---
--- This transaction can come from any era, old (Byron) and new (Babbage).
-data Tx where
-    Tx
-        :: forall era
-        . (Show (TxEra era), Api.IsCardanoEra era)
-        => Api.CardanoEra era
-        -> TxEra era
-        -> Tx
-
-deriving instance Show Tx
-
-instance Eq Tx where
-    (==) (Tx ByronEra x) = \case
-        Tx ByronEra y -> x == y
-        _ -> False
-    (==) (Tx ShelleyEra x) = \case
-        Tx ShelleyEra y -> x == y
-        _ -> False
-    (==) (Tx AllegraEra x) = \case
-        Tx AllegraEra y -> x == y
-        _ -> False
-    (==) (Tx MaryEra x) = \case
-        Tx MaryEra y -> x == y
-        _ -> False
-    (==) (Tx AlonzoEra x) = \case
-        Tx AlonzoEra y -> x == y
-        _ -> False
-    (==) (Tx BabbageEra x) = \case
-        Tx BabbageEra y -> x == y
-        _ -> False
-
-{- Note [SeeminglyRedundantPatternMatches]
-
-When writing code that handles multiple eras, it is highly recommended
-that you pattern match on each era invidiually and handle the specialized
-types that arise.
-
-When doing these pattern matches, you may find that some cases
-may seem highly redundant.
-However, the types of the patterns are wildly different due to
-type-level programming, so the cases are not actually redundant.
-
-In other words, it is best to keep each case separate,
-even if this seemingly duplicates code,
-and *not* attempt to refactor them into common function,
-as the type signatures for those functions are more complicated
-than any simplification that would result from avoided repetition.
-
-Look, you have been warned!
--}
-
-{-------------------------------------------------------------------------------
-    TxId
--------------------------------------------------------------------------------}
-type TxId = W.Hash "Tx"
-
--- | Compute the 'txId' of a transaction by hashing the transaction body.
-computeTxId :: Tx -> TxId
-computeTxId (Tx era tx) = case era of
```

Comment:
Uh, can we keep this function around — though probably as `computeTxId :: EraFun Tx (K TxId)`? 🤔
It's definitely a "feature" of a transaction. Eventually, I'd like to (re-)export all the "features" from `Cardano.Wallet.Types.Read.Tx`, as this is the abstraction boundary for "all things related to a transaction".

Reply:
Yes we can but not here.  In my plan that is an `EraFun Tx Id` and will go with the extractions . Also in my design there is a section dedicated to Primitives where there will be an `EraFun Id (K (W.Hash "Tx"))` which is what we had here. So in particular, Read.Tx module should not import any primitive data.